tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+use crate::loom::alloc::Track;
+use crate::loom::cell::CausalCheck;
+use crate::task::waker::waker_ref;
+use crate::task::{raw, Error, Header, Schedule, Snapshot, Task, Trailer};
+
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem::{ManuallyDrop, MaybeUninit};
+use std::pin::Pin;
+use std::ptr::{self, NonNull};
+use std::task::{Context, Poll, Waker};
+
+/// Typed raw task handle
+pub(super) struct Harness<T, S: 'static> {
+    header: NonNull<Header<S>>,
+    trailer: NonNull<Trailer>,
+    future_or_output: *mut (),
```

Comment:
Is this optimization measurable? It seems quite complex vs using an `enum`.

Reply:
Thanks for questioning this. Your comment + a number of questions on reddit motivated me to recheck my assumptions.

Turns out, I once again fell for premature optimization and thought that I could do better than the compiler.

It turns out that using `Box::new` for task allocation + storing the future / output in an `enum` is faster and includes less unsafe code!

Here are some bechmarks:

# `std::alloc` / union strategy

### System allocator

```
test chained_spawn ... bench:     170,149 ns/iter (+/- 31,140)
test ping_pong     ... bench:     502,818 ns/iter (+/- 60,887)
test spawn_many    ... bench:   7,376,991 ns/iter (+/- 463,349)
test yield_many    ... bench:  15,753,987 ns/iter (+/- 741,801)
```
```
test chained_spawn ... bench:     174,004 ns/iter (+/- 7,921)
test ping_pong     ... bench:     557,004 ns/iter (+/- 21,449)
test spawn_many    ... bench:   7,083,360 ns/iter (+/- 391,059)
test yield_many    ... bench:  16,090,016 ns/iter (+/- 573,387)
```
```
test chained_spawn ... bench:     171,500 ns/iter (+/- 23,582)
test ping_pong     ... bench:     554,794 ns/iter (+/- 18,205)
test spawn_many    ... bench:   7,527,440 ns/iter (+/- 137,863)
test yield_many    ... bench:  16,497,012 ns/iter (+/- 259,349)
```
```
test chained_spawn ... bench:     169,776 ns/iter (+/- 10,908)
test ping_pong     ... bench:     519,973 ns/iter (+/- 8,821)
test spawn_many    ... bench:   7,529,021 ns/iter (+/- 296,098)
test yield_many    ... bench:  17,024,497 ns/iter (+/- 597,971)
```

### jemalloc

```
test chained_spawn ... bench:     135,688 ns/iter (+/- 9,588)
test ping_pong     ... bench:     320,315 ns/iter (+/- 41,349)
test spawn_many    ... bench:   6,906,364 ns/iter (+/- 616,085)
test yield_many    ... bench:  10,298,150 ns/iter (+/- 1,630,291)
```
```
test chained_spawn ... bench:     132,957 ns/iter (+/- 11,469)
test ping_pong     ... bench:     314,489 ns/iter (+/- 27,316)
test spawn_many    ... bench:   6,775,606 ns/iter (+/- 478,108)
test yield_many    ... bench:  10,446,986 ns/iter (+/- 721,916)
```
```
test chained_spawn ... bench:     137,228 ns/iter (+/- 32,515)
test ping_pong     ... bench:     322,023 ns/iter (+/- 70,225)
test spawn_many    ... bench:   6,668,900 ns/iter (+/- 307,068)
test yield_many    ... bench:  10,190,525 ns/iter (+/- 537,933)
```
```
test chained_spawn ... bench:     133,574 ns/iter (+/- 16,976)
test ping_pong     ... bench:     332,107 ns/iter (+/- 38,208)
test spawn_many    ... bench:   6,681,921 ns/iter (+/- 258,790)
test yield_many    ... bench:  10,090,287 ns/iter (+/- 1,875,526)
```

# `Box::new` / `enum`

### System allocator

```
test chained_spawn ... bench:     147,943 ns/iter (+/- 6,673)
test ping_pong     ... bench:     537,744 ns/iter (+/- 20,928)
test spawn_many    ... bench:   7,454,898 ns/iter (+/- 283,449)
test yield_many    ... bench:  16,771,113 ns/iter (+/- 733,424)
```
```
test chained_spawn ... bench:     148,070 ns/iter (+/- 18,396)
test ping_pong     ... bench:     563,103 ns/iter (+/- 26,209)
test spawn_many    ... bench:   7,629,426 ns/iter (+/- 132,804)
test yield_many    ... bench:  17,628,425 ns/iter (+/- 433,587)
```
```
test chained_spawn ... bench:     147,166 ns/iter (+/- 7,167)
test ping_pong     ... bench:     484,848 ns/iter (+/- 21,424)
test spawn_many    ... bench:   7,145,832 ns/iter (+/- 338,560)
test yield_many    ... bench:  12,792,609 ns/iter (+/- 861,848)
```
```
test chained_spawn ... bench:     144,589 ns/iter (+/- 23,384)
test ping_pong     ... bench:     530,946 ns/iter (+/- 16,992)
test spawn_many    ... bench:   7,279,071 ns/iter (+/- 513,115)
test yield_many    ... bench:  15,741,148 ns/iter (+/- 687,794)
```
```
test chained_spawn ... bench:     144,472 ns/iter (+/- 6,304)
test ping_pong     ... bench:     515,494 ns/iter (+/- 11,031)
test spawn_many    ... bench:   7,408,886 ns/iter (+/- 322,611)
test yield_many    ... bench:  16,777,689 ns/iter (+/- 387,922)
```

### jemalloc

```
test chained_spawn ... bench:     102,045 ns/iter (+/- 4,255)
test ping_pong     ... bench:     297,896 ns/iter (+/- 48,701)
test spawn_many    ... bench:   6,203,292 ns/iter (+/- 235,247)
test yield_many    ... bench:   9,607,488 ns/iter (+/- 662,828)
```
```
test chained_spawn ... bench:     100,797 ns/iter (+/- 37,800)
test ping_pong     ... bench:     258,828 ns/iter (+/- 72,306)
test spawn_many    ... bench:   6,122,506 ns/iter (+/- 121,452)
test yield_many    ... bench:   9,759,652 ns/iter (+/- 711,285)
```
```
test chained_spawn ... bench:     100,797 ns/iter (+/- 37,800)
test ping_pong     ... bench:     258,828 ns/iter (+/- 72,306)
test spawn_many    ... bench:   6,122,506 ns/iter (+/- 121,452)
test yield_many    ... bench:   9,759,652 ns/iter (+/- 711,285)
```
```
test chained_spawn ... bench:     103,094 ns/iter (+/- 40,394)
test ping_pong     ... bench:     290,606 ns/iter (+/- 34,244)
test spawn_many    ... bench:   6,166,750 ns/iter (+/- 100,825)
test yield_many    ... bench:   9,584,838 ns/iter (+/- 397,545)
```